source/slang/slang-emit-wgsl.cpp

```diff
@@ -0,0 +1,1013 @@
+        SLANG_UNEXPECTED("16 bit integer value emitted");
+        return;
+    case kIROp_Int64Type:
+    case kIROp_IntPtrType:
+        m_writer->emit("i64");
+        return;
+    case kIROp_UIntPtrType:
+        m_writer->emit("u64");
+        return;
+    case kIROp_StructType:
+        m_writer->emit(getName(type));
+        return;
+
+    case kIROp_VectorType:
+    {
+        auto vecType = (IRVectorType*)type;
+        emitVectorTypeNameImpl(
+            vecType->getElementType(), getIntVal(vecType->getElementCount())
+        );
+        return;
+    }
+    case kIROp_MatrixType:
+    {
+        auto matType = (IRMatrixType*)type;
+        // We map matrices in Slang to WGSL matrices that represent the transpose.
+        // (See note on "terminology reversal".)
+        emitTransposedMatrixType(matType->getElementType(), matType);
+        return;
+    }
+    case kIROp_SamplerStateType:
+    {
+        m_writer->emit("sampler");
+        return;
+    }
+
+    case kIROp_SamplerComparisonStateType:
+    {
+        m_writer->emit("sampler_comparison");
+        return;
+    }
+
+    case kIROp_PtrType:
+    case kIROp_InOutType:
+    case kIROp_OutType:
+    case kIROp_RefType:
+    case kIROp_ConstRefType:
+    {
+        auto ptrType = cast<IRPtrTypeBase>(type);
+        m_writer->emit("ptr<");
+        emit((AddressSpace)ptrType->getAddressSpace());
+        m_writer->emit(", ");
+        emitType((IRType*)ptrType->getValueType());
+        m_writer->emit(">");
+        return;
+    }
+
+    case kIROp_ArrayType:
+    {
+        m_writer->emit("array<");
+        emitType((IRType*)type->getOperand(0));
+        m_writer->emit(", ");
+        emitVal(type->getOperand(1), getInfo(EmitOp::General));
+        m_writer->emit(">");
+        return;
+    }
+    default:
+        break;
+
+    }
+
+}
+
+void WGSLSourceEmitter::emitVarAttributesImpl(IRType * type, IRVarLayout* layout)
+{
+
+    for (auto attr : layout->getOffsetAttrs())
+    {
+
+        // TODO:
+        // This is not correct. For the moment this is just here as a hack to make
+        // @binding and @group unique, so that we can pass WGSL compile tests.
+        // This will have to be revisited when we actually want to supply resources to
+        // shaders.
+        if (
+            (type->getOp() == kIROp_HLSLRWStructuredBufferType) ||
+            (type->getOp() == kIROp_HLSLStructuredBufferType)
+            )
+        {
+            m_writer->emit("@binding(");
+            m_writer->emit(attr->getOffset());
+            m_writer->emit(") ");
+            m_writer->emit("@group(");
+            m_writer->emit(attr->getSpace());
+            m_writer->emit(") ");
+
+            return;
+        }
+    }
+
+}
+
+void WGSLSourceEmitter::emitVarKeywordImpl(IRType * type, bool const isConstant)
+{
+    if (isConstant)
+        m_writer->emit("const");
+    else
+        m_writer->emit("var");
+    if (type->getOp() == kIROp_HLSLRWStructuredBufferType)
```

Comment:
It feels wrong to test for structured buffer types during emit logic of a `var`. Looks like this needs to be lowered into a variable of pointer type during legalization.

Reply:
> It feels wrong to test for structured buffer types during emit logic of a `var`. 

I agree. This was a temporary hack that I had meant to clean up...

> Looks like this needs to be lowered into a variable of pointer type during legalization.

The equivalent of `RWStructuredBuffer<int> gOutputBuffer;` in WGSL is `var<storage, read_write> gOutputBuffer : array<i32>;` so it can't be a pointer type in this one case when we declare it.

Everywhere else, though, I have the hack described in `WGSLSourceEmitter::emitOperandImpl` to change all places where `gOutputBuffer` is used as an operand to use instead `&gOutputBuffer`.

Reply:
So I can lower structured buffer types to pointer types, but then I somehow need to deal with the case where structured buffer globals are declared, because those storage buffer declarations are not supposed to be pointers in WGSL.

That does sound preferable to me, but I'd like to know what you think before fixing it this way.

Reply:
> In this case, we should make sure an IRGlobalParam of StructuredBuffer is in the form of `IRGlobalParam p : Ptr<Array<E>, readwrite_storage>`, and insert `OpLoad` at use sites of the global param during legalization. That way we won't need any specialized logic for structured buffers as they don't exist when we get to emit.

So the `OpLoad` at the usage sites would take a `ptr<storage, array<E>, read_write>` and give us `array<E>`?

I think most often in he WGSL code, the correct type for `buffer` that references a `StorageBuffer<E>` is actually `ptr<uniform, array<E>, read_only>`. For examle if you pass `buffer` to function, WGSL will force you to pass a ptr.
...so I don't think we'd want `OpLoad` to "dereference" and give us `array<E>` in any case that I know of.

Reply:
After all the legalization step, you shouldn't be seeing buffers being passed around through function calls. They will be legalized into global variables, and all functions will directly access that global variable, instead of load it and pass it as argument to another function.

Back to this problem, I think the solution is:
1. We still lower structured buffer types to `PTR<Array<T>>` during legalization.
2. Structured buffer elements will be accessed with the `kIROp_StructuredBufferLoad` instruction, we should keep these instructions as is during legalization.
3.  Make sure `kIROp_StructuredBufferLoad`, `IRRWStructuredBufferGetElementPtr` is handled correctly, so that we emit `buffer[0]` instead of `&buffer[0]`.

Reply:
> After all the legalization step, you shouldn't be seeing buffers being passed around through function calls. They will be legalized into global variables, and all functions will directly access that global variable, instead of load it and pass it as argument to another function.

To clarify, are you saying I'm missing an _existing_ legalization step like this, or to I need to implement a legalization step like this?

If it already exists, could you give me a pointer? I'd like to figure out why it's seemingly not being called.

Reply:
oh, wait, this is to make `storage, read_write` appear after `var`.
then yeah this is a reasonable place for the logic after all.
sorry for the distraction, i think we are good with what we have.